### PR TITLE
Chromium: Implement basic WebXR input support

### PR DESCRIPTION
**Pull request by zakharvoit**
This change implements basic WebXR input support, including controller position, orientation and pressed buttons.

This was tested on Oculus Quest 2 controllers.

### CODE REVIEW

File: wolvic/browser/vr/wvr_manager.cc [hunk 1]
```diff
@@ -373,14 +373,14 @@ WvrManager::GetInputSourceState() {
...
-    //     continue;
+    if (controller.type == mozilla::gfx::VRControllerType::_empty)
+      continue;
 
     device::mojom::XRInputSourceStatePtr input_source =
         device::mojom::XRInputSourceState::New();
 
-    input_source->source_id = i;
+    // The source_id == 0 is not supported, so we're using a 1-based id here.
+    input_source->source_id = i + 1;
```
Comment (svillar) [inline wolvic/browser/vr/wvr_manager.cc:422]: I was about to upload a patch for this too :smile: 

For better understanding I think we can copy exactly the same comment that is in the openxr_input_helper.cc file

File: wolvic/browser/vr/wvr_manager.cc [hunk 2]
```diff
@@ -403,15 +403,24 @@ WvrManager::GetInputSourceState() {
...
 
-    gfx::Transform input_from_pointer;
-    WvrMatToTransform(controller.axisValue, &input_from_pointer);
-    input_source->description->input_from_pointer = input_from_pointer;
+    input_source->gamepad = device::Gamepad();
+    input_source->gamepad->buttons_length = controller.numButtons;
+    input_source->gamepad->hand =
+        controller.hand == mozilla::gfx::ControllerHand::Left
+            ? device::GamepadHand::kLeft
+            : device::GamepadHand::kRight;
```
Comment (svillar) [inline wolvic/browser/vr/wvr_manager.cc]: This is repited a few lines above. We should just initialize a variable and use it in both places.

**Review by svillar — CHANGES_REQUESTED** [review]
Much better now. However the aim is not correct, it's displaced in both position and orientation

Apart from that I asked for 2 minor nits

File: wolvic/browser/vr/wvr_manager.cc (hunk 2 above)
Comment (jdapena) [reply wolvic/browser/vr/wvr_manager.cc]: For this kind of conversions between mozilla and chrome structures, I would like to see explicit convert methods (that can be inlined in the future if we worry about performance).

Something like:

device::GamepadHand ToGamepadHand(mozilla::gfx::ControllerHand moz_hand) {
  switch(moz_hand) {
    ...
}

Speciall if this is used from several files. I don't want to have the explicit evaluation everywhere.

Same with other conversions we are inlining in this method, making it too much complex. I would prefer helper methods for each of the mozilla to chrome structure conversions (i.e. handedness, targetRayMode).

In those methods we can also have specific warnings for the case a new value comes from Mozilla side we don't know about.

File: wolvic/browser/vr/wvr_manager.cc [hunk 3]
```diff
@@ -403,15 +403,24 @@ WvrManager::GetInputSourceState() {
...
+    for (uint32_t j = 0; j < controller.numButtons; ++j) {
```
Comment (jdapena) [inline wolvic/browser/vr/wvr_manager.cc]: We should make sure we don't pass the kButtonsLengthCap (and show a warning if that happens).

File: wolvic/browser/vr/wvr_manager.cc (hunk 2 above)
Comment (zakharvoit) [reply wolvic/browser/vr/wvr_manager.cc]: I refactored this conversion using the suggestions from @jdapena. However, the conversion above is using XRHandedness instead of device::GamepadHand so this will have to be a separate method. I didn't touch the rest of the code in this change but I will submit a separate PR refactoring the rest of the conversion code in a similar way.

File: wolvic/browser/vr/wvr_manager.cc (hunk 3 above)
Comment (zakharvoit) [reply wolvic/browser/vr/wvr_manager.cc]: Done in ToGamepadHand()

**Comment by svillar** [comment]
Refactorings are fine, but we need to get the aim/grip poses right before landing.

**Review by jdapena — APPROVED** [review]

**Review by svillar — CHANGES_REQUESTED** [review]
Are the aim/grip poses issues fixed? I haven't seen any change related to that, just refactorings

**Comment by zakharvoit** [comment]
@svillar Sorry for a delay, I proposed a slightly different approach (on the corresponding gitlab issue) with extending the moz_external_vr.h to include an additonal field that we already have on the Wolvic side. I'll create a draft change for Wolvic implementing this today and update this change to make use of that new field. Please refer to the gitlab issue for more explanation.

**Comment by zakharvoit** [comment]
> Are the aim/grip poses issues fixed? I haven't seen any change related to that, just refactorings

Should be fixed now with the latest commit, depends on the following PR https://github.com/Igalia/wolvic/pull/849

**Review by jdapena — APPROVED** [review]

File: wolvic/browser/vr/moz_external_vr.h [hunk 4]
```diff
@@ -153,6 +151,7 @@ enum class VRControllerType : uint8_t {
   PicoGaze,
   PicoG2,
   PicoNeo2,
+  Pico4,
```
Comment (svillar) [inline wolvic/browser/vr/moz_external_vr.h:156]: Unrelated change, please do it in a separate commit/PR

File: wolvic/browser/vr/moz_external_vr.h [hunk 5]
```diff
@@ -429,6 +428,11 @@ struct VRControllerState {
   // https://immersive-web.github.io/webxr/#dom-xrinputsource-targetrayspace
   VRPose targetRayPose;
 
+  // When Cap_Orientation is set in flags, targetRayOrientation corresponds
+  // to the orientation of the target ray (origin of the ray is determined by
+  // the controller position). Only used by the Chromium backend.
+  float targetRayOrientation[4];
```
Comment (svillar) [inline wolvic/browser/vr/moz_external_vr.h]: As mentioned I don't think this is the right approach. We should already have all the available data because we can do it with Gecko. Have you tried to mimic Gecko's code to get the info you need?

**Review by svillar — CHANGES_REQUESTED** [review]
I don't think this is the right approach
